src/overlays/actors/ovl_Bg_Treemouth/z_bg_treemouth.c

```diff
@@ -21,29 +37,215 @@ const ActorInit Bg_Treemouth_InitVars = {
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Treemouth/func_808BCAF0.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Treemouth/func_808BCB8C.s")
+static InitChainEntry sInitChain[] = {
+    ICHAIN_U8(unk_1F, 5, ICHAIN_CONTINUE),     ICHAIN_VEC3F(scale, 1, ICHAIN_CONTINUE),
+    ICHAIN_F32(unk_F4, 8000, ICHAIN_CONTINUE), ICHAIN_F32(unk_F8, 300, ICHAIN_CONTINUE),
+    ICHAIN_F32(unk_FC, 300, ICHAIN_STOP),
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Treemouth/BgTreemouth_Update.s")
+// unused
+f32 D_808BD9C4[] = {
+    -2746.0f, 545.0f, 4694.0f, -2654.0f, 146.0f, 4534.0f,
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Treemouth/BgTreemouth_Draw.s")
+extern Gfx D_060009D0[];
+extern UNK_TYPE D_06000E94;
+
+void BgTreemouth_SetupAction(BgTreemouth* this, BgTreemouthActionFunc actionFunc) {
+    this->actionFunc = actionFunc;
+}
+
+void BgTreemouth_Init(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    BgTreemouth* this = THIS;
+    u32 localC = 0;
+
+    Actor_ProcessInitChain(thisx, sInitChain);
+    DynaPolyInfo_SetActorMove(&this->dyna, 0);
+    DynaPolyInfo_Alloc(&D_06000E94, &localC);
+    this->dyna.dynaPolyId = DynaPolyInfo_RegisterActor(globalCtx, &globalCtx->colCtx.dyna, thisx, localC);
+    ActorShape_Init(&thisx->shape, 0.0f, NULL, 0.0f);
+    Actor_SetHeight(thisx, 50.0f);
+
+    if ((gSaveContext.sceneSetupIndex < 4) && (LINK_IS_CHILD)) {
+        BgTreemouth_SetupAction(this, func_808BC8B8);
+    } else {
+        if ((LINK_IS_ADULT) || (gSaveContext.sceneSetupIndex == 7)) {
+            this->unk_168 = 0.0f;
+            BgTreemouth_SetupAction(this, BgTreemouth_DoNothing);
+        } else {
+            this->unk_168 = 1.0f;
+            BgTreemouth_SetupAction(this, func_808BC6F8);
+        }
+    }
+
+    thisx->textId = 0x905;
+}
+
+void BgTreemouth_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    BgTreemouth* this = THIS;
+
+    DynaPolyInfo_Free(globalCtx, &globalCtx->colCtx.dyna, this->dyna.dynaPolyId);
+}
+
+void func_808BC65C(BgTreemouth* this, GlobalContext* globalCtx) {
+    CsCmdActorAction* npcAction;
+
+    if ((globalCtx->csCtx.state != 0)) {
+        npcAction = globalCtx->csCtx.npcActions[0];
+        if (npcAction != NULL) {
+            if (npcAction->action == 2) {
+                BgTreemouth_SetupAction(this, func_808BC80C);
+            } else if (npcAction->action == 3) {
+                Audio_PlaySoundGeneral(NA_SE_EV_WOODDOOR_OPEN, &D_801333D4, 4, &D_801333E0, &D_801333E0, &D_801333E8);
+                BgTreemouth_SetupAction(this, func_808BC6F8);
+            }
+        }
+    }
+}
+
+void func_808BC6F8(BgTreemouth* this, GlobalContext* globalCtx) {
+    Vec3f sp34;
+
+    if (this->unk_168 < 1.0f) {
+        this->unk_168 += 0.01f;
+    } else {
+        this->unk_168 = 1.0f;
+    }
+
+    if ((gSaveContext.sceneSetupIndex == 6) && (globalCtx->csCtx.frames >= 0x2BD) && (!(globalCtx->state.frames & 7))) {
```

Comment:
Frame values are usually better as decimal, and this `!(frames & 7)` isn't a boolean check, so it would be better as `(frames % 8 == 0)` imo.